code/controllers/subsystem/sounds.dm

```diff
@@ -0,0 +1,91 @@
+	/// Assoc list, "[channel]" = either the datum using it or TRUE for an unsafe-reserved (datumless reservation) channel
+	var/list/using_channels = list()
+	/// Assoc list datum = list(channel1, channel2, ...) for what channels something reserved.
+	var/list/using_channels_by_datum = list()
+	/// List of all available channels with associations set to TRUE for fast lookups/allocation.
+	var/list/available_channels
+
+/datum/controller/subsystem/sounds/Initialize()
+	setup_available_channels()
+	return ..()
+
+/datum/controller/subsystem/sounds/proc/setup_available_channels()
+	available_channels = list()
+	for(var/i in 1 to using_channels_max)
+		available_channels[num2text(i)] = TRUE
+
+/// Removes a channel from using list.
+/datum/controller/subsystem/sounds/proc/free_sound_channel(channel)
+	channel = num2text(channel)
+	var/using = using_channels[channel]
+	using_channels -= channel
+	if(using)
+		using_channels_by_datum[using] -= channel
+		if(!length(using_channels_by_datum[using]))
+			using_channels_by_datum -= using
+	available_channels[channel] = TRUE
+
+/// Frees all the channels a datum is using.
+/datum/controller/subsystem/sounds/proc/free_datum_channels(datum/D)
+	var/list/L = using_channels_by_datum[D]
+	if(!L)
+		return
+	for(var/channel in L)
+		using_channels -= channel
+		available_channels[channel] = TRUE
+	using_channels_by_datum -= D
+
+/// Frees all datumless channels
+/datum/controller/subsystem/sounds/proc/free_datumless_channels()
+	free_datum_channels(DATUMLESS)
+
+/// NO AUTOMATIC CLEANUP - If you use this, you better manually free it later! Returns an integer for channel.
+/datum/controller/subsystem/sounds/proc/reserve_sound_channel_datumless()
+	var/channel = random_available_channel_text()
+	if(!channel)		//oh no..
+		return FALSE
+	available_channels -= channel
+	using_channels[channel] = DATUMLESS
+	LAZYINITLIST(using_channels_by_datum[DATUMLESS])
+	using_channels_by_datum[DATUMLESS] += channel
+	return text2num(channel)
+
+/// Reserves a channel for a datum. Automatic cleanup only when the datum is deleted. Returns an integer for channel.
+/datum/controller/subsystem/sounds/proc/reserve_sound_channel(datum/D)
+	if(!D)		//i don't like typechecks but someone will fuck it up
+		CRASH("Attempted to reserve sound channel without datum using the managed proc.")
+	var/channel = random_available_channel_text()
+	if(!channel)
+		return FALSE
+	available_channels -= channel
+	using_channels[channel] = D
+	LAZYINITLIST(using_channels_by_datum[D])
+	using_channels_by_datum[D] += channel
+	return text2num(channel)
+
+/// Random available channel, returns text.
+/datum/controller/subsystem/sounds/proc/random_available_channel_text()
+	return pick(available_channels)
+
+/// Random available channel, returns number
+/datum/controller/subsystem/sounds/proc/random_available_channel()
+	return text2num(pick(available_channels))
```

Comment:
I disagree with how you implemented this.
Imagine this datastructure
This is how it looks on init. i = 1 and n = array.len.
The value of each cell is its index.
![image](https://user-images.githubusercontent.com/47324920/83878092-e4b1fb80-a73b-11ea-96ac-9f7c414f80cb.png)

Reserving a channel does the following. It decrements n by 1, if i is now greater than n, it will be set to 1. It returns the value n used to point to, in this case ``d``
![image](https://user-images.githubusercontent.com/47324920/83878360-57bb7200-a73c-11ea-8242-d021400426dd.png)

Getting the next channel does this. It increments i by 1, if i now greater than n, i will be set to 1. It returns the value i used to point to, in this case ``a``.
![image](https://user-images.githubusercontent.com/47324920/83878807-f942c380-a73c-11ea-8fae-807c00536cc5.png)

This is all trivial, now to the real meat of the structure. The function ``free``. Free takes an x integer greater than n, increments n and swaps the values in list[n] and list[x]
![image](https://user-images.githubusercontent.com/47324920/83879351-ca791d00-a73d-11ea-9d9d-1c37dcc14297.png)
, now this has one problem, the values in the list are no longer ordered.
If we executed this
```
L = init()
L.reserve()
L.reserve()
L.free(4)
```
L would have the value [1,2,4,3]. This means if something were now to naively free 3, it would throw an error as position and value are not actually dependent on each other.
This is pretty easy to fix, just have a list that uses reserved channels as the key and the position in the array as a value.
Update this list whenever values get swapped in the array.

What is the benefit of this: No random sound conflicts.
Downsides: Requires slightly more memery in the worst case. Probably less performant because we don't have arrays
 If this were C, this solution would literally have constant complexity, unlike pick which is linear.
God I hate byond.